cpp/src/cylon/net/ops/gather.cpp

```diff
@@ -0,0 +1,154 @@
+                     const int gather_root,
+                     const bool gather_from_root,
+                     std::shared_ptr<cylon::Allocator> allocator,
+                     std::unique_ptr<int32_t []> & all_buffer_sizes,
+                     std::vector<std::shared_ptr<cylon::Buffer>> & receive_buffers,
+                     std::vector<std::vector<int32_t>> & displacements,
+                     std::shared_ptr<cylon::CylonContext> ctx
+                     ){
+
+    // first gather table buffer sizes
+    std::vector<int32_t> local_buffer_sizes;
+    if(AmIRoot(gather_root, ctx) && !gather_from_root) {
+        local_buffer_sizes = serializer->getEmptyTableBufferSizes();
+    } else {
+        local_buffer_sizes = serializer->getBufferSizes();
+    }
+
+    int32_t num_buffers = local_buffer_sizes.size();
+
+    // gather size buffers
+    if (AmIRoot(gather_root, ctx)) {
+        all_buffer_sizes = std::make_unique<int32_t []>(ctx->GetWorldSize() * num_buffers);
+    }
+
+    int status = MPI_Gather(local_buffer_sizes.data(),
+                            num_buffers,
+                            MPI_INT32_T,
+                            all_buffer_sizes.get(),
+                            num_buffers,
+                            MPI_INT32_T,
+                            gather_root,
+                            MPI_COMM_WORLD);
+    if (status != MPI_SUCCESS) {
+        return cylon::Status(cylon::Code::ExecutionError, "MPI_Gather failed!");
+    }
+
+    std::vector<int32_t> total_buffer_sizes;
+    if (AmIRoot(gather_root, ctx)) {
+        totalBufferSizes(all_buffer_sizes.get(), num_buffers, ctx->GetWorldSize()).swap(total_buffer_sizes);
+    }
+
+    auto requests = std::make_unique<MPI_Request []>(num_buffers);
+    auto statuses = std::make_unique<MPI_Status []>(num_buffers);
+    std::vector<uint8_t *> send_buffers = serializer->getDataBuffers();
+
+    for (int32_t i = 0; i < num_buffers; ++i) {
+        if(AmIRoot(gather_root, ctx)) {
+            std::shared_ptr<cylon::Buffer> receive_buf;
+            allocator->Allocate(total_buffer_sizes[i], &receive_buf);
+            std::vector<int32_t> receive_counts = receiveCounts(all_buffer_sizes.get(),
+                                                                i,
+                                                                num_buffers,
+                                                                ctx->GetWorldSize());
+            std::vector<int32_t> disp_per_buffer = displacementsPerBuffer(all_buffer_sizes.get(),
+                                                                          i,
+                                                                          num_buffers,
+                                                                          ctx->GetWorldSize());
+            displacements.push_back(disp_per_buffer);
+
+            status = MPI_Igatherv(send_buffers.at(i),
```

Comment:
let's use `send_buffers[i]`. `vector::at(i)` checks i against the bounds (which is generally slow)

Reply:
done.